File: mobile/modules/screen-lock/android/src/main/java/expo/modules/screenlock/ScreenLockModule.kt
package expo.modules.screenlock

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * ScreenLock — reports whether the screen turned off (device locked) since the
 * last `reset()`, so a focus reminder fires only when the user switched to
 * another app, not when they locked the phone to study.
 *
 * Android is reliable here: ACTION_SCREEN_OFF fires on lock/screen-off, while
 * an app-switch leaves the screen on. The receiver is registered dynamically
 * (screen state broadcasts can't be declared in the manifest).
 */
class ScreenLockModule : Module() {
  private var lockedSinceReset = false
  private var receiver: BroadcastReceiver? = null

  private val appContextOrNull: Context?
    get() = appContext.reactContext?.applicationContext

  override fun definition() = ModuleDefinition {
    Name("ScreenLock")

    Function("reset") {
      lockedSinceReset = false
    }

    Function("consumeLocked") {
      val value = lockedSinceReset
      lockedSinceReset = false
      value
    }

    OnCreate {
      val ctx = appContextOrNull ?: return@OnCreate
      val r = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
          if (intent?.action == Intent.ACTION_SCREEN_OFF) {
            lockedSinceReset = true
          }
        }
      }
      receiver = r
      val filter = IntentFilter(Intent.ACTION_SCREEN_OFF)
      if (Build.VERSION.SDK_INT >= 34) {
        ctx.registerReceiver(r, filter, Context.RECEIVER_NOT_EXPORTED)
      } else {
        ctx.registerReceiver(r, filter)
      }
    }

    OnDestroy {
      receiver?.let {
        try {
          appContextOrNull?.unregisterReceiver(it)
        } catch (_: Exception) {
        }
      }
      receiver = null
    }
  }
}

File: mobile/modules/screen-lock/ios/ScreenLockModule.swift
import ExpoModulesCore
import UIKit

/**
 * ScreenLock — reports whether the device screen was locked since the last
 * `reset()`. Used so a focus reminder fires only when the user switched to
 * ANOTHER APP, not when they simply locked the phone to study.
 *
 * iOS has no reliable, App-Store-safe "screen locked" event, so we use the
 * public `protectedDataWillBecomeUnavailable` signal (fires when a passcode-
 * protected device locks). It can be delayed or, on passcode-less devices, not
 * fire at all — hence this is best-effort. The JS layer fails safe: if a lock
 * can't be confirmed it errs toward "locked" so it never nags on a real lock.
 */
public final class ScreenLockModule: Module {
  private var lockedSinceReset = false
  private var observer: NSObjectProtocol?

  public func definition() -> ModuleDefinition {
    Name("ScreenLock")

    // Clear the flag at the start of an absence (called when the app backgrounds
    // during a running session).
    Function("reset") {
      self.lockedSinceReset = false
    }

    // Read-and-clear: did the screen lock during the last absence?
    Function("consumeLocked") { () -> Bool in
      let value = self.lockedSinceReset
      self.lockedSinceReset = false
      return value
    }

    OnCreate {
      self.observer = NotificationCenter.default.addObserver(
        forName: UIApplication.protectedDataWillBecomeUnavailableNotification,
        object: nil,
        queue: nil
      ) { [weak self] _ in
        self?.lockedSinceReset = true
      }
    }

    OnDestroy {
      if let observer = self.observer {
        NotificationCenter.default.removeObserver(observer)
        self.observer = nil
      }
    }
  }
}
